### PR TITLE
completed assessment

### PR DESCRIPTION
**Pull request by Hieuhuule**
# Submitting Week 2 Assessment

### ✅ Completed Files

Select the following as confirmation that the files in this repository are complete.

- [x ] Coding Challenges
- [ x] Coding Concepts
- [x ] Interview Questions

### 📝 Feedback

If you would like specific feedback on a particular problem please leave details here:

### CODE REVIEW

File: code-challenges.test.js [hunk 1]
```diff
@@ -1,5 +1,8 @@
 // ASSESSMENT 2: Coding Practical Questions with Jest
 
+// const { it } = require("node:test")
+// const { describe } = require("yargs")
+
```
Comment (SunkissedQueen) [inline code-challenges.test.js:5]: the invasion of the yargs

File: code-challenges.test.js [hunk 2]
```diff
@@ -36,10 +39,32 @@ const object3 = { number: -7 }
...
-// b) Create the function that makes the test pass.
+// PSUEDO CODE
+// describe names the function
+// it describes what the function does
+// expect gives an expected outcome given a specific parameter
+// create a function that takes an object as an arguement
+// create a conditional statement within the function to determine if the value assigned to the number key is divisable by 3
+// return a string statement if number is divisible by 3 or not
 
+describe("divisibleByThree", () => {
```
Comment (SunkissedQueen) [inline code-challenges.test.js:50]: nice test

File: code-challenges.test.js [hunk 3]
```diff
@@ -36,10 +39,32 @@ const object3 = { number: -7 }
...
+        expect(divisibleByThree(object2)).toEqual("0 is divisible by three")
+        expect(divisibleByThree(object3)).toEqual("-7 is not divisible by three")
+    })
+})
+
+// b) Create the function that makes the test pass.
 
 
+const divisibleByThree = (obj) => {
+    if (obj.number % 3 === 0) {
```
Comment (SunkissedQueen) [inline code-challenges.test.js:62]: yes to abstracting the value using dot notation

File: code-challenges.test.js [hunk 4]
```diff
@@ -50,9 +75,28 @@ const randomNouns2 = ["temperature", "database", "chopsticks", "mango", "deducti
...
+// create a function that takes in an array of strings
+// use map() to iterate over the array
+// each iteration uses toUpperCase() method on first character and concats the remainder of the word, adding .toLowerCase() makes sure the rest of the work is grammatically correct
+// map returns the new array with the first letter capitalized
+
+const upCaseFirst = (arr) => {
+    return arr.map(element => {
+        return element.charAt(0).toUpperCase() + element.slice(1).toLowerCase();
+    })
+}
```
Comment (SunkissedQueen) [inline code-challenges.test.js:97]: like that you consider the edge cases of arguments besides the test variables

File: code-challenges.test.js [hunk 5]
```diff
@@ -50,9 +75,28 @@ const randomNouns2 = ["temperature", "database", "chopsticks", "mango", "deducti
...
 
+    // This one was a doozy, I couldn't figure it out using what we learned and I had to look for a solution online. I had two major roadblocks on this problem, the first was that I tried tunneling down to the first char and .toUpperCase() it, but I wasn't getting the return I was expecting. I spent a significant amount of time on it before I discovered strings were immutable. The second road block was using slice. We hadn't practiced it at all, so trying to wrap my mind around it was pretty difficult.
```
Comment (SunkissedQueen) [inline code-challenges.test.js:99]: good job using the concept of learning that we showcase. As you have seen in your research, there are endless methods and tools to aid with logic. Very difficult to cover everything. The key is to remember what datatype you are using and that there are built-in methods available for your use.

File: code-challenges.test.js [hunk 6]
```diff
@@ -65,5 +109,30 @@ const vowelTester2 = "academy"
...
+    for (let i = 0; i < str.length; i++) {
+        if (str[i] === "a" || str[i] === "A" || 
+            str[i] === "i" || str[i] === "I" ||
+            str[i] === "e" || str[i] === "E" ||
+            str[i] === "o" || str[i] === "O" ||
+            str[i] === "u" || str[i] === "U") {
+            return str.indexOf(str[i])
+        } 
+    }
+}
```
Comment (SunkissedQueen) [inline code-challenges.test.js:138]: mission accomplished

File: coding-concepts.js [hunk 7]
```diff
@@ -10,8 +10,8 @@
 const cohort = "Echo 2022"
 // console.log(cohort.split(""))
 
-// a) Your answer:
-// b) Verify and explain:
+// a) Your answer: An array of every character.
+// b) Verify and explain: My guess was correct, the .split() method operates on an string and breaks up every character according the the parameter, in this case ("").
```
Comment (SunkissedQueen) [inline coding-concepts.js:14]: good

File: coding-concepts.js [hunk 8]
```diff
@@ -21,26 +21,27 @@ const greeter = (name) => {
...
 
 // --------------------3) What will this log?
 
 const multipliedByTwo = [4, 5, 6, 7, 8].map(number => number * 2)
 // console.log(multipliedByTwo)
 
-// a) Your answer:
-// b) Verify and explain:
+// a) Your answer: [8, 10, 12, 14, 16]
+// b) Verify and explain: correct. map iterates over the original array and returns a new array and then multiplied every element inside of it by 2, the new array is then assigned to the variable multipliedByTwo
```
Comment (SunkissedQueen) [inline coding-concepts.js:35]: yes

File: coding-concepts.js [hunk 9]
```diff
@@ -21,26 +21,27 @@ const greeter = (name) => {
...
 
 // --------------------4) What will this log?
 
 const onlyOdds = [11, 12, 13, 14, 15].filter(number => number % 2 !== 0)
 // console.log(onlyOdds)
 
-// a) Your answer:
-// b) Verify and explain:
+// a) Your answer: [11, 13, 15]
+// b) Verify and explain: correct. filter returns an array containing every number that is not evenly divisible by 2.
```
Comment (SunkissedQueen) [inline coding-concepts.js:44]: yes

File: coding-concepts.js [hunk 10]
```diff
@@ -53,8 +54,8 @@ const myCodingSkills = {
 }
 // console.log(myCodingSkills.languages[0])
 
-// a) Your answer:
-// b) Verify and explain:
+// a) Your answer: "JavaScript"
+// b) Verify and explain: correct. It uses dot and bracket notation to tunnel down to the language key of myCodingSkills and then accessed the first item of the array located there.
```
Comment (SunkissedQueen) [inline coding-concepts.js:58]: correct

File: coding-concepts.js [hunk 11]
```diff
@@ -69,5 +70,5 @@ class Learn {
 const learnStudent = new Learn("George")
 // console.log(learnStudent)
 
-// a) Your answer:
-// b) Verify and explain:
+// a) Your answer: it will create a copy of the Learn class and assign it to the learnStudent variable with "George" as the value of the key student.
+// b) Verify and explain: correct. A class is a blueprint for objects. By calling new on the class name and providing it with an argument that matches it's constructor, you create a new unique instance of that class.
```
Comment (SunkissedQueen) [inline coding-concepts.js:74]: exactly

File: interview-questions.md [hunk 12]
```diff
@@ -8,33 +8,37 @@ Then, research the question to expand on your answer. Even if you feel you have
 
 1. What is the difference between a parameter and an argument?
 
-Your answer:
+Your answer: A parameter and argument are the same, I've heard them used interchangeably as what goes in the parenthesis for a method or function.
 
-Researched answer:
+Researched answer: A parameter is a variable in the function definition. It is a placeholder and doesn't have a set value. An argument is a set value that is passed during function invocation. So when a function isn't given a specific set of data, what is in the parenthesis is a called a parameter (basically a placeholder). But if the function is passed specified information, then it's called an argument. A good way to visualize a parameter is as default information filled in a search bar prior to you typing stuff in, once you start typing, it's an argument.
```
Comment (SunkissedQueen) [inline interview-questions.md:13]: nice

File: interview-questions.md [hunk 13]
```diff
@@ -8,33 +8,37 @@ Then, research the question to expand on your answer. Even if you feel you have
...
 
-Your answer:
+Your answer: .map() takes in the array and a parameter for an element, then within the code block, you can write code that affects the element, then that code is iterated over each element, and finally .map() returns a new array with the affected code. If you wanted to keep the information, you would have to assign the new array into a variable.
 
-Researched answer:
+Researched answer: The .map() method can take in 4 parameters: 
+    currentValue - the value of the current element, REQUIRED;
+    index - the index of the current element, OPTIONAL;
+    arr - the array of the current element, OPTIONAL;
+    thisValue - A value passed to the function to be used as its this value, OPTIONAL. Default value undefined.
```
Comment (SunkissedQueen) [inline interview-questions.md:23]: good research. suggest showing an example with the `thisValue` argument

File: interview-questions.md [hunk 14]
```diff
@@ -8,33 +8,37 @@ Then, research the question to expand on your answer. Even if you feel you have
...
+Your answer: .map() takes in the array and a parameter for an element, then within the code block, you can write code that affects the element, then that code is iterated over each element, and finally .map() returns a new array with the affected code. If you wanted to keep the information, you would have to assign the new array into a variable.
 
-Researched answer:
+Researched answer: The .map() method can take in 4 parameters: 
+    currentValue - the value of the current element, REQUIRED;
+    index - the index of the current element, OPTIONAL;
+    arr - the array of the current element, OPTIONAL;
+    thisValue - A value passed to the function to be used as its this value, OPTIONAL. Default value undefined. 
 
 3. What is the difference between map and filter?
```
Comment (SunkissedQueen) [inline interview-questions.md:26]: suggest adding more content with examples

File: interview-questions.md [hunk 15]
```diff
@@ -8,33 +8,37 @@ Then, research the question to expand on your answer. Even if you feel you have
...
-Researched answer:
+Researched answer: .map() is defined above, so I'll describe the filter method. It's a method that creates a new array filled with elements that pass a test provided by a function. It doesn't change the original array.
 
 4. What is the DOM?
 
-Your answer:
+Your answer: Hmm, I don't know, but let's go to the researched answer.
 
-Researched answer:
+Researched answer: Document Object Model (DOM), is a platform and language-neutral interface that allows programs and scripts to dynamically access and update the content, structure, and style of a document. It's separated into 3 different parts, Core DOM, XML DOM, and HTML DOM, which are all standard models for their type. It's the standard for how to get, change, add, or delete elements in their type model. That's a long winded book answer, but it's very simply how things are laid out on a webpage and how to interact with them.
```
Comment (SunkissedQueen) [inline interview-questions.md:35]: this is some interesting research. Suggest a brief explanation of Core DOM, XML DOM, HTML DOM

File: interview-questions.md [hunk 16]
```diff
@@ -8,33 +8,37 @@ Then, research the question to expand on your answer. Even if you feel you have
...
 
 5. What is React? Why would you use it?
 
-Your answer:
+Your answer: React is a language created by Facebook which allows you to build websites and applications that can scale from a small number of users to billions of users, all by making code modular. It's language is based on JavaScript and is familiar to anyone experienced with HTML and CSS. Besides in very specific situations, the JSX language operates the same as JavaScript.
 
-Researched answer:
+Researched answer: React is a free and open-source frontend JavaScript library for building user interfaes based on UI components. It is mained by Meta (formerly Facebook) and a community of individual developers and companies. It's aim is to allow developers to easily create fast user interfaces for websites and applications alike. The main concept of React.js is virtual DOM. It is a tree based on JavaScript components created with React that mimics a DOM tree. It does the least amount of DOM manipulation possible in order to keep your React compoenents up to date.
```
Comment (SunkissedQueen) [inline interview-questions.md:41]: nice

File: interview-questions.md [hunk 17]
```diff
@@ -8,33 +8,37 @@ Then, research the question to expand on your answer. Even if you feel you have
...
```
Comment (SunkissedQueen) [inline interview-questions.md:42]: The Stretch is not required. Just some nice knowledge to have. However, you do need to provide definitions for the looking ahead terms.

**Review by SunkissedQueen — CHANGES_REQUESTED** [review]
Please provide definitions for the `Looking Ahead: Terms for Next Week`

**Review by SunkissedQueen — APPROVED** [review]
Hieu, thanks for the update. Better job on the interview questions. Please maintain that approach with defining the topic and showing how you understand the material. I would say also consider including some projects that used the topic being discussed. For the coding challenges, I see that you brought in some high order functions. Let's continue introducing other methods that will help reduce the amount of lines for coding. Definitely get a function that makes the test pass then try to find another technique to solve the problem with less lines.
